doc/cprover-manual/function-contracts.md

```diff
@@ -0,0 +1,502 @@
+   `__CPROVER_ensures`, and/or `__CPROVER_assigns`;
+2. Compile the C program using `goto-cc` to generate a single binary, e.g.,
+   `goto-cc -o main.goto *.c`;
+3. Enforce contract *c* to function *f* by instrument the code using
+   `goto-instrument` with the flag `--enforce-contract <fun>`, where `<fun>` is
+the signature for function *f*, e.g., `goto-instrument --enforce-contract
+add_overflow main.goto main-with-contracts.goto`;
+4. Verify the resulting binary with CBMC, e.g., `cbmc main-with-contracts.goto`.
+
+To apply the same process to all functions with contracts at once, use the flag
+`--enforce-all-contracts`.
+
+
+### Replacement
+
+Assuming a function contract *c* is a sound characterization of the behavior of
+the function *f*, CBMC will replace a call to *f* with *c* as follows:
+
+1. Adding assertions for all preconditions specified in the `__CPROVER_requires`
+   clauses; 
+2. Adding non-deterministic assignments for each symbol listed in the
+   `__CPROVER_assigns` clause;
+3. Assuming all postconditions described in the `__CPROVER_ensures` clauses;
+
+In our example, consider that a function `foo` may call `add_overflow`.
+
+```c
+void foo()
+{
+  uint32_t a;
+  uint32_t b;
+  uint32_t *out = malloc(sizeof(*out));
+
+  int rval = add_overflow(a, b, out);
+  
+  assert((rval == 0) ==> (*out == a + b));
+  /* ... */
+}
+```
+
+The call to `add_overflow` function is replaced by CBMC as follows
+
+```c
+void foo()
+{
+  uint32_t a;
+  uint32_t b;
+  uint32_t *out = malloc(sizeof(*out));
+
+  /* Function Contract Replacement */
+  __CPROVER_assert(out != NULL, "Check requires clause");
+  *out = nondet_uint32_t();
+  int return_value_add_overflow = nondet_int();
+  __CPROVER_assume(return_value_add_overflow == SUCCESS || return_value_add_overflow == FAILURE);
+  __CPROVER_assume((return_value_add_overflow == SUCCESS) ==> (*out == (a + b)));
+  
+  int rval = return_value_add_overflow;
+  assert((rval == 0) ==> (*out == a + b));
+  /* ... */
+}
+```
+
+This replacement happens at the GOTO level. Thus, to replace all calls to
+function *f* with its contracts, follow these instructions:
+
+1. Annotate the function *f* with contracts using `__CPROVER_requires`,
+   `__CPROVER_ensures`, and/or `__CPROVER_assigns`;
+2. Compile the C program using `goto-cc` to generate a single binary, e.g.,
+   `goto-cc -o main.goto *.c`;
+3. Replace all calls to function *f* with its contract by instrument the code
+   using `goto-instrument` with the flag `--replace-call-with-contract <fun>`,
+where `<fun>` is the signature for function *f*, e.g., `goto-instrument
+--replace-call-with-contract add_overflow main.goto main-with-contracts.goto`;
+4. Verify the resulting binary with CBMC, e.g., `cbmc main-with-contracts.goto`.
+
+To apply the same process to all functions with contracts at once, use the flag
+`--replace-all-calls-with-contracts`.
+
+## Built-in Constructs
+
+### Historic variables
```

Comment:
Historic -> [History](https://rosettacode.org/wiki/History_variables)

Reply:
Fixed it.